p2p/src/msg.rs

```diff
@@ -228,19 +251,50 @@ impl Writeable for MsgHeader {
 	}
 }
 
-impl Readable for MsgHeader {
-	fn read(reader: &mut dyn Reader) -> Result<MsgHeader, ser::Error> {
+impl Readable for MsgHeaderWrapper {
+	fn read(reader: &mut dyn Reader) -> Result<MsgHeaderWrapper, ser::Error> {
 		let m = magic();
 		reader.expect_u8(m[0])?;
 		reader.expect_u8(m[1])?;
-		let (t, len) = ser_multiread!(reader, read_u8, read_u64);
+
+		// Read the msg header.
+		// We do not yet know if the msg type is one we support locally.
+		let (t, msg_len) = ser_multiread!(reader, read_u8, read_u64);
+
+		// Attempt to convert the msg type byte into one of our known msg type enum variants.
+		// Check the msg_len while we are at it.
 		match Type::from_u8(t) {
-			Some(ty) => Ok(MsgHeader {
-				magic: m,
-				msg_type: ty,
-				msg_len: len,
-			}),
-			None => Err(ser::Error::CorruptedData),
+			Some(msg_type) => {
+				// TODO 4x the limits for now to leave ourselves space to change things.
+				let max_len = max_msg_size(msg_type) * 4;
+				if msg_len > max_len {
+					error!(
+						"Too large read {:?}, max_len: {}, msg_len: {}.",
+						msg_type, max_len, msg_len
+					);
+					return Err(ser::Error::TooLargeReadErr);
+				}
+
+				Ok(MsgHeaderWrapper::Known(MsgHeader {
+					magic: m,
+					msg_type,
+					msg_len,
+				}))
+			}
+			None => {
+				// Unknown msg type, but we still want to limit how big the msg is.
+				// Default to max_block_size (4x as above for space to change things).
+				let max_len = max_block_size() * 4;
```

Comment:
why `max_block_size`?

Reply:
Ahh good question - `max_block_size()` is the largest msg size we support in `max_msg_size()`.

We should probably pull this out into something more meaningful like `default_max_msg_size()`
and document this better.